httpcore/_synchronization.py

```diff
@@ -112,117 +144,110 @@ def __exit__(
+        if (semaphore := self._semaphore) is None:
+            semaphore = self.setup()
+        await semaphore.acquire()
 
     async def release(self) -> None:
-        if self._backend == "trio":
-            self._trio_semaphore.release()
-        elif self._backend == "asyncio":
-            self._anyio_semaphore.release()
+        cast(_LockProto, self._semaphore).release()
+
+
+T = TypeVar("T")
 
 
 class AsyncShieldCancellation:
     # For certain portions of our codebase where we're dealing with
     # closing connections during exception handling we want to shield
     # the operation from being cancelled.
     #
-    # with AsyncShieldCancellation():
-    #     ... # clean-up operations, shielded from cancellation.
-
-    def __init__(self) -> None:
-        """
-        Detect if we're running under 'asyncio' or 'trio' and create
-        a shielded scope with the correct implementation.
-        """
-        self._backend = current_async_library()
-
-        if self._backend == "trio":
-            self._trio_shield = trio.CancelScope(shield=True)
-        elif self._backend == "asyncio":
-            self._anyio_shield = anyio.CancelScope(shield=True)
-
-    def __enter__(self) -> "AsyncShieldCancellation":
-        if self._backend == "trio":
-            self._trio_shield.__enter__()
-        elif self._backend == "asyncio":
-            self._anyio_shield.__enter__()
-        return self
-
-    def __exit__(
-        self,
-        exc_type: Optional[Type[BaseException]] = None,
-        exc_value: Optional[BaseException] = None,
-        traceback: Optional[TracebackType] = None,
-    ) -> None:
-        if self._backend == "trio":
-            self._trio_shield.__exit__(exc_type, exc_value, traceback)
-        elif self._backend == "asyncio":
-            self._anyio_shield.__exit__(exc_type, exc_value, traceback)
+    # async def cleanup():
+    #     ...
+    #
+    # # clean-up operations, shielded from cancellation.
+    # await AsyncShieldCancellation.shield(cleanup)
+
+    @staticmethod
+    async def shield(shielded: Callable[[], Coroutine[Any, Any, None]]) -> None:
+        if current_async_backend() == "trio":
+            with trio.CancelScope(shield=True):
+                await shielded()
+        elif sys.version_info < (3, 11):  # pragma: nocover
+            with anyio_shield(shield=True):
+                await shielded()
+        else:
+            await AsyncShieldCancellation._asyncio_shield(shielded)  # pragma: nocover
+
+    @staticmethod
+    async def _asyncio_shield(
+        shielded: Callable[[], Coroutine[Any, Any, None]],
+    ) -> None:  # pragma: nocover
+        inner_task = asyncio.create_task(shielded())
+        try:
+            await asyncio.shield(inner_task)
+        except (asyncio.exceptions.CancelledError, asyncio.CancelledError):
+            # Let the inner_task to complete as it was shielded from the cancellation
+            await inner_task
```

Comment:
No need instantiation, so we can convert to function then:
```suggestion
async def async_cancel_shield(shielded: Callable[[], Coroutine[Any, Any, None]]) -> None:
    if current_async_backend() == "trio":
        with trio.CancelScope(shield=True):
            await shielded()
    elif sys.version_info < (3, 11):  # pragma: nocover
        with anyio_shield(shield=True):
            await shielded()
    else:
        inner_task = asyncio.create_task(shielded())
        try:
            await asyncio.shield(inner_task)
        except (asyncio.exceptions.CancelledError, asyncio.CancelledError):
            # Let the inner_task to complete as it was shielded from the cancellation
            await inner_task
```

Reply:
Done, also needed to add the function to the `unasync` utility.